### ISSUE TITLE
How to use encrypted template file

### ISSUE THREAD

**Issue by KevinNitroG** [twpayne/chezmoi#4170]
## What exactly are you trying to do?

I want to have `rclone.conf` encrypted in `/home/.chezmoitemplates/rclone/encrypted_rclone.conf.asc` and be encrypted in `~\AppData\Roaming\rclone\rclone.conf` on Windows, and `dot_config/rclone/rclone.conf` on Linux

## What have you tried so far?

> [!NOTE]
>
> Config in chezmoi repo

- At `/home/.chezmoitemplates/rclone/encrypted_rclone.conf.asc`
  ```
  -----BEGIN PGP MESSAGE-----
  ...
  -----END PGP MESSAGE-----
  ```
- At `home/AppData/Roaming/rclone/rclone.conf.tmpl`
  ```
  {{ template "rclone/encrypted_rclone.conf.asc" . }}
  ```

I haven't config for Linux. Also I already removed it when os not equal Windows.

## Where else have you checked for solutions?

* [x] I have read [chezmoi's user guide](https://chezmoi.io/user-guide/command-overview/), and not found the answer.
* [x] I have searched [chezmoi's reference guide](https://chezmoi.io/reference/), and not found the answer.
* [ ] Other, please give details.

## Output of any commands you've tried with `--verbose` flag

```console
$ chezmoi --verbose apply C:\Users\kevinnitro\AppData\Roaming\rclone\rclone.conf
```

the content of the rclone.conf file is the raw encrypted file.

> [!IMPORTANT]
>
> Nothing showed up to the terminal

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
```

```
RESULT    CHECK                       MESSAGE
ok        version                     v2.56.0, commit 3558b71471e91230ca4e87c1969c9dca62cd9b54, built at 2024-12-14T20:15:02Z, built by goreleaser
warning   latest-version              v2.57.0
ok        os-arch                     windows/amd64
ok        systeminfo                  Microsoft Windows 11 Home (10.0.22631 N/A Build 22631)
ok        go-version                  go1.23.4 (gc)
ok        executable                  ~/scoop/apps/chezmoi/current/chezmoi.exe
ok        upgrade-method              replace-executable
ok        config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-12-31T23:53:13+07:00
warning   source-dir                  ~/.local/share/chezmoi is a git working tree (dirty)
ok        suspicious-entries          no suspicious entries
warning   working-tree                ~/.local/share/chezmoi is a git working tree (dirty)
ok        dest-dir                    ~ is a directory
ok        cd-command                  found C:/Windows/system32/cmd.exe
ok        cd-args                     cmd
ok        diff-command                found ~/scoop/shims/delta.exe
ok        edit-command                found ~/scoop/shims/nvim.exe
ok        edit-args                   nvim
ok        git-command                 found ~/scoop/shims/git.exe, version 2.47.0
ok        merge-command               found ~/scoop/shims/nvim.exe
ok        shell-command               found C:/Windows/system32/cmd.exe
ok        shell-args                  'C:\\Windows\\system32\\cmd.exe'
info      age-command                 age not found in $PATH
ok        gpg-command                 found ~/scoop/apps/git/current/usr/bin/gpg.exe, version 2.4.5
info      pinentry-command            not set
info      1password-command           op not found in $PATH
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
info      keepassxc-command           keepassxc-cli not found in $PATH
info      keepassxc-db                not set
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set
```

</details>

**Comment by halostatue** [comment]
I don't think that this will work, but please try `includeTemplate` instead of `template`. You may also want to pipe the result through `| decrypt` (see the example at [decrypt](https://www.chezmoi.io/reference/templates/functions/decrypt/)).

**Comment by KevinNitroG** [comment]
> I don't think that this will work, but please try `includeTemplate` instead of `template`. You may also want to pipe the result through `| decrypt` (see the example at [decrypt](https://www.chezmoi.io/reference/templates/functions/decrypt/)).

I'm gonna try it tomorrow and give the feedback. But I wonder will it be able to sync back from home dir to work dir (chezmoi repo)?

**Comment by halostatue** [comment]
In general, encrypted files should be used either as:

- sources of data (`filename | include | decrypt | fromJson`) for use in templates
- files to be decrypted in place
- files to be included in a template without modification (as the example suggests)

I would be surprised if chezmoi tried decrypting a template before executing it. It probably *could*, but I don't know that would be something that should be enabled and encouraged.

**Comment by KevinNitroG** [comment]
@halostatue I have configured like this

- `\home\.chezmoitemplates\rclone\encrypted_rclone.conf.asc` contains encrypted data
- `home\AppData\Roaming\rclone\rclone.conf.tmpl` whose content:
  ```
  {{ includeTemplate "rclone/encrypted_rclone.conf.asc" | decrypt }}
  ```

It works with chezmoi apply. However, I will have to edit raw rclone.conf, encrypt myself, and put it into chezmoitemplates folder manually. Is there any workflow better than this?

**Comment by KevinNitroG** [comment]
I wonder can we use chezmoi run on change feature to achieve this? I'm quite new to chezmoi's some features and currently I have not much knowledge on chezmoi run action...

**Comment by halostatue** [comment]
> @halostatue I have configured like this
> 
> * `\home\.chezmoitemplates\rclone\encrypted_rclone.conf.asc` contains encrypted data
> * `home\AppData\Roaming\rclone\rclone.conf.tmpl` whose content:
>   ```
>   {{ includeTemplate "rclone/encrypted_rclone.conf.asc" | decrypt }}
>   ```
> 
> It works with chezmoi apply. However, I will have to edit raw rclone.conf, encrypt myself, and put it into chezmoitemplates folder manually. Is there any workflow better than this?

IMO, there are several workflows better than this, but none of them work any better if you want to keep the entire configuration encrypted in repo.

- https://github.com/search?q=path%3Arclone.conf.tmpl&type=code (this includes some Docker template stuff, but still appears to be mostly chezmoi)
- https://github.com/search?q=path%3Arclone.conf.tmpl+chezmoi&type=code (exclusively chezmoi, all using per-field config)

I personally use 1Password for this sort of thing (composing blobs) but have a couple of different ways of approaching it:

- Managing `~/.aws/credentials` and `~/.aws/config` as collections of blobs:
  - https://github.com/halostatue/dotfiles/blob/main/home/private_dot_aws/private_credentials.tmpl
  - https://github.com/halostatue/dotfiles/blob/main/home/private_dot_aws/private_config.tmpl

- Managing `~/.config/git/identities/*` with appropriate references for SSH keys in `~/.ssh/config` and `~/.config/git/config`
  - https://github.com/halostatue/dotfiles/blob/main/home/private_dot_config/git/identities/run_onchange_git_identities.sh.tmpl
  - https://github.com/halostatue/dotfiles/blob/main/home/private_dot_ssh/private_config.tmpl
  - https://github.com/halostatue/dotfiles/blob/main/home/.chezmoitemplates/1password_git_identities.tmpl
  - https://github.com/halostatue/dotfiles/blob/main/home/private_dot_config/git/config.tmpl

**Comment by KevinNitroG** [comment]
Thank you. I'm gonna have a look into these for references. I'm closing this issue now.
